### PR TITLE
[mono][jit] Internal memcpy uses SIMD

### PR DESCRIPTION
**Pull request by jandupej**
Copying large enough structs byval is now done in SIMD if alignment allows or CPU does not mind unaligned access.

### CODE REVIEW

File: src/mono/mono/mini/memory-access.c [hunk 1]
```diff
@@ -130,6 +130,12 @@ mini_emit_memcpy (MonoCompile *cfg, int destreg, int doffset, int srcreg, int so
 	}
 
 	g_assert (align > 0);
+	
+#ifndef NO_UNALIGNED_ACCESS	
+	// Unaligned access is not a problem on this platform. Use the longest
+	// possible vector.
+	goto copy_16;
```
Comment (vargaz) [inline src/mono/mono/mini/memory-access.c:137]: Note that the cpu allowing unaligned access does not mean that its going to be fast.

Comment (jandupej) [reply src/mono/mono/mini/memory-access.c:137]: What I'm trying to do is optimize memcpying types like `Matrix4x4` which have proper alignment, but do not register as SIMD types. Other than checking class name against a known list, do you see a way of detecting those?

Comment (vargaz) [reply src/mono/mono/mini/memory-access.c:137]: The function has an align argument which is supposed to be the guaranteed alignment of src/dest.

**Comment by jandupej** [comment]
It appears that allowing unaligned writes could break ordering if the struct sits squarely on a cache line boundary, which we cannot prevent. Closing.

**Pull request closed without merging**